### DISCUSSION TITLE
Getting RevocationCheckingPolicy() takes no arguments

### DISCUSSION THREAD

**Discussion by talha-abdulrab** [MatthiasValvekens/pyHanko#118] [Q&A]
Hi!
I am using html2pdf for generating pdf and it utilizes pyHanko. I am not aware of depth implementation but I am getting this error from pyHanko
RevocationCheckingPolicy() takes no arguments
DeveloperExtention() takes no arguments
can you please help me out.

**Comment by MatthiasValvekens** [comment]
Hi, can you post a full stack trace? This might be a version conflict between `pyhanko` and `pyhanko-certvalidator`, so version numbers would also help.

**Reply by talha-abdulrab** [reply]
File "/var/www/dinController/venv3.10/lib/python3.10/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/var/www/dinController/venv3.10/lib/python3.10/site-packages/django/core/handlers/base.py" in _get_response
  115.                 response = self.process_exception_by_middleware(e, request)

File "/var/www/dinController/venv3.10/lib/python3.10/site-packages/django/core/handlers/base.py" in _get_response
  113.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/var/www/dinController/apps/host/views.py" in vms_list_pdf
  407.     return util.render_to_pdf(template, context)

File "/var/www/dinController/lib/utils.py" in render_to_pdf
  1042.         from xhtml2pdf import pisa

File "/var/www/dinController/venv3.10/lib/python3.10/site-packages/xhtml2pdf/pisa.py" in <module>
  26. from xhtml2pdf.document import pisaDocument

File "/var/www/dinController/venv3.10/lib/python3.10/site-packages/xhtml2pdf/document.py" in <module>
  24. from xhtml2pdf.builders.signs import PDFSignature

File "/var/www/dinController/venv3.10/lib/python3.10/site-packages/xhtml2pdf/builders/signs.py" in <module>
  6. from pyhanko.sign import signers, timestamps

File "/var/www/dinController/venv3.10/lib/python3.10/site-packages/pyhanko/sign/__init__.py" in <module>
  1. from .signers import *

File "/var/www/dinController/venv3.10/lib/python3.10/site-packages/pyhanko/sign/signers/__init__.py" in <module>
  10. from .constants import (

File "/var/www/dinController/venv3.10/lib/python3.10/site-packages/pyhanko/sign/signers/constants.py" in <module>
  52. ESIC_EXTENSION_1 = DeveloperExtension(

Exception Type: TypeError at /hs/vms/pdf/628/
Exception Value: DeveloperExtension() takes no arguments

**Reply by talha-abdulrab** [reply]
Versions
pyHanko                  0.13.1
pyhanko-certvalidator    0.19.5

**Reply by talha-abdulrab** [reply]
My problem is this is only not working in production, its working fine on my local environment.

**Comment by MatthiasValvekens** [comment]
Hi, @talha-abdulrab, this is most likely due to an outdated version of Python on your prod server. There's a bug in some of the prior 3.10 releases of CPython that (among other things) causes the `@dataclass` decorator to break in subinterpreters. See the discussion [on this `mod_wsgi` issue](https://github.com/GrahamDumpleton/mod_wsgi/issues/729) for more information.

Upgrading your Python version to the most recent release (or downgrading to 3.9) should fix the problem.

**Comment by talha-abdulrab** [comment]
@MatthiasValvekens  thank you so much for your help, I'll try updating python to recent version.

**Comment by DanielHefti** [comment]
I had the same error after upgrading my server to Python 3.12. It seems that pyHanko-certvalidator isn't compatible with Python 3.12 yet. After downgrading to python 3.11 again, everything works fine.

Python 3.12
pyHanko 0.23.2
pyHanko certvalidator 0.26.3

**Reply by MatthiasValvekens** [reply]
That would be extremely strange--pyHanko CI has included 3.12 since forever. Granted, I forgot to include 3.12 in the test matrix for pyhanko-certvalidator, but after rectifying that, the tests pass just fine on 3.12 as well.

Remember that this error means that the creation of the dataclass _type_ object itself is failing. That's not really a thing that happens in pure Python code; it has to be a bug in CPython itself, or in some C extension.

In fact, there appears to be a regression of the 3.10 issue I linked above in Python 3.12: https://github.com/python/cpython/issues/110279. This should've been fixed in 3.12.1. What's the exact release of Python 3.12 you're running?

**Reply by DanielHefti** [reply]
I've tried using Python 3.12.0. That's the latest version available in the cloud environment. Perhaps the problem is related to CentOs 7.9.2009 and the installed gcc version 4.8.5. I had similar problems with the uWsgi package.

**Reply by MatthiasValvekens** [reply]
No, the bug was only fixed in 3.12.1, so the fact that it fails in 3.12.0 is not entirely unexpected :)

**Reply by MatthiasValvekens** [reply]
That said, CentOS 7.9.2009 is also an almost-prehistoric distro. I'm actually surprised that anyone is even packaging 3.12 for a distribution that old ;).
